mongodbatlas/fw_resource_mongodbatlas_database_user.go

```diff
@@ -0,0 +1,531 @@
+package mongodbatlas
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+
+	"github.com/hashicorp/terraform-plugin-framework-validators/stringvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/attr"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/planmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringdefault"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema/stringplanmodifier"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	matlas "go.mongodb.org/atlas/mongodbatlas"
+)
+
+const (
+	databaseUserResourceName = "database_user"
+)
+
+var _ resource.Resource = &DatabaseUserRS{}
+var _ resource.ResourceWithImportState = &DatabaseUserRS{}
+
+type tfDatabaseUserModel struct {
+	ID               types.String `tfsdk:"id"`
+	ProjectID        types.String `tfsdk:"project_id"`
+	AuthDatabaseName types.String `tfsdk:"auth_database_name"`
+	Username         types.String `tfsdk:"username"`
+	Password         types.String `tfsdk:"password"`
+	X509Type         types.String `tfsdk:"x509_type"`
+	OIDCAuthType     types.String `tfsdk:"oidc_auth_type"`
+	LDAPAuthType     types.String `tfsdk:"ldap_auth_type"`
+	AWSIAMType       types.String `tfsdk:"aws_iam_type"`
+	Roles            types.Set    `tfsdk:"roles"`
+	Labels           types.Set    `tfsdk:"labels"`
+	Scopes           types.Set    `tfsdk:"scopes"`
```

Comment:
I have found that replacing types.Set for the actual slice, example []tfRoleModel, greatly reduces boilerplate code for setting and getting the slice values, as well as having to define each corresponding ObjectType